src/QsCompiler/Transformations/CallGraph.cs

```diff
@@ -0,0 +1,569 @@
+        {
+            if (target == null) throw new ArgumentNullException(nameof(target));
+            var success = TryCombineTypeResolutions(out combined, resolutions);
+            combined = combined.Where(kvp => kvp.Key.Item1.Equals(target)).ToImmutableDictionary();
+            return success;
+        }
+
+        public static bool TryCombineTypeResolutions(out TypeParameterResolutions combined, params TypeParameterResolutions[] resolutionDictionaries)
+        {
+            if (!resolutionDictionaries.Any())
+            {
+                combined = TypeParameterResolutions.Empty;
+                return true;
+            }
+
+            var combinedBuilder = ImmutableDictionary.CreateBuilder<Tuple<QsQualifiedName, NonNullable<string>>, ResolvedType>();
+            var success = true;
+
+            static Tuple<QsQualifiedName, NonNullable<string>> AsTypeResolutionKey(QsTypeParameter tp) => Tuple.Create(tp.Origin, tp.TypeName);
+            static bool IsSelfResolution(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                res.Resolution is ResolvedTypeKind.TypeParameter tp && tp.Item.Origin.Equals(typeParam.Item1) && tp.Item.TypeName.Equals(typeParam.Item2);
+            static bool IsConstrictiveResolution(Tuple<QsQualifiedName, NonNullable<string>> typeParam, ResolvedType res) =>
+                res.Resolution is ResolvedTypeKind.TypeParameter tp && typeParam.Item1.Equals(tp.Item.Origin) && typeParam.Item2.Value != tp.Item.TypeName.Value;
+
+            foreach (var resolutionDictionary in resolutionDictionaries)
+            {
+                // Contains a lookup of all the keys in the combined resolutions whose value needs to be updated
+                // if a certain type parameter is resolved by the currently processed dictionary.
+                var mayBeReplaced = combinedBuilder
+                    .Where(kv => kv.Value.Resolution.IsTypeParameter)
+                    .ToLookup(
+                        kv => AsTypeResolutionKey(((ResolvedTypeKind.TypeParameter)kv.Value.Resolution).Item),
+                        entry => entry.Key);
+
+                // Do any replacements for type parameters that may be replaced with values in the current dictionary.
+                // This needs to be done first to cover an edge case.
+                foreach (var (typeParam, paramRes) in resolutionDictionary.Where(entry => mayBeReplaced.Contains(entry.Key)))
+                {
+                    // Get all the parameters whose value is dependent on the current resolution's type parameter,
+                    // and update their values with this resolution's value.
+                    foreach (var keyInCombined in mayBeReplaced[typeParam])
+                    {
+                        // Check that we are not doing an excessively constrictive resolution.
```

Comment:
```suggestion
                        // Check that we are not doing a resolution that constricts what values a type parameter can have.
```

Reply:
Hmm, that wording is confusing to me because all resolutions constrict the type of a type parameter. But I suppose you made the suggestions because the original wording was confusing to you? I had based the original wording off of the what I remembered the error message was for this scenario in other languages, but I can't seem to find this error message now. Could be that I am misremembering.

Reply:
The excessive didn't mean much to me (aside from that I already know what it does). 
maybe something along the lines of "constricts multiple type parameters"?

Reply:
Reworded the comment for clarity.